x-pack/elastic-agent/pkg/core/plugin/app/status.go

```diff
@@ -0,0 +1,71 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package app
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sanathkr/go-yaml"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/server"
+)
+
+// ApplicationStatusHandler expects that only Application is registered in the server and updates the
+// current state of the application from the OnStatusChange callback from inside the server.
+//
+// In the case that an application is reported as failed by the server it will then restart the application, unless
+// it expects that the application should be stopping.
+type ApplicationStatusHandler struct{}
+
+// OnStatusChange is the handler called by the GRPC server code.
+//
+// It updates the status of the application and handles restarting the application is needed.
+func (*ApplicationStatusHandler) OnStatusChange(state *server.ApplicationState, status proto.StateObserved_Status, msg string) {
+	app, ok := state.App().(*Application)
+	if !ok {
+		panic(errors.New("only *Application can be registered when using the ApplicationStatusHandler", errors.TypeUnexpected))
```

Comment:
do we need to crash agent here?

Reply:
i think we should, because using this handler means that only `*Application` is registered in the server and if not then we did something wrong in the code

I would rather make it breaking all of agent so we can fix, then allowing it to hide the issue and be harder to debug in the future